### ISSUE TITLE
Feedback on Marl

### ISSUE THREAD

**Issue by JeremyBYU** [google/marl#116]
Hi,
I'm just going to leave some feeback and thoughts on `marl` as recommended from this [external issue](https://github.com/JeremyBYU/TaskLibraryBenchmark/issues/1). Note that my experiences with `marl` are limited and involve only the incomplete [benchmark repository](https://github.com/JeremyBYU/TaskLibraryBenchmark) I made and a point cloud processing library (for robotics), [polylidar](https://github.com/JeremyBYU/polylidar), which is having a major refactor being done which also includes `marl`.

> Oh nice! How are you finding marl²?

I find marl to be very simple to *build* and *integrate* into my existing project. The CMake build integration is very important to me and seems to be setup correctly such it was trivial to add to my project. Also the `Scheduler` and `WaitGroup` primitives seem simple enough and it was apparent how to integrate them into `polylidar`. Overall I enjoy how simple it is too describe work in `tasks` and describe dependencies with the wait groups.

> Given that your benchmarks put marl mostly in last place in terms of performance, can I ask what convinced you to go with marl³?

I chose to try out `marl` _before_ those benchmarks were completed : ) .  But the main reasons were the following:

1. Many of the tasks for `polylidar` were dynamic and nested,  meaning you have no idea how many subtasks must be completed inside of  a task (and all those subtasks must finish before their parent task is completed). It seemed like marl was primarily focused on this scenario. Also I do not have thousands of tasks generally, maybe like 100 total. I figured the difference between all these libraries might be small at such a low amount of tasks.
2. I saw that `marl` had hand-written assembly for the arm architecture for its [fiber/thread scheduling](https://github.com/google/marl/blob/master/src/osfiber_asm_aarch64.S). I will be running some of my code on an RPI4 and felt more comfortable using a library that had tested with that architecture. Of course that's just a way of saying I was too lazy to (initially) benchmark performance between architectures and implementations and was relying upon googles name and heavy use of android/arm to *assume* good testing and performance. Note I do plan to run my benchmark repo on my rpi4 after it has solidified better.

So that lead me to just experiment with marl first. And it worked. Very well. 

However I do have some general comments:

My system:
CPU:  Ryzen 3900X (12 Core/24 Thread)
RAM: 32 GB
OS: Ubuntu 18.04
Compiler: C++ 14 GCC 7.something

1. There seems to be a 300 *micro*second penalty loss when using marl, even if only 1 thread is used. You can see where I create the scheduler [here](https://github.com/JeremyBYU/polylidar/blob/4a291b58743c7879f39fc779c487e77778bfaa91/src/Polylidar/Polylidar.cpp#L120).

I have benchmarks that compared using marl and not using marl for a light workload that only requires about 300 microseconds of work (because). In this output `Optimized` means using `marl`. 

```
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/1/real_time_mean                   336 us          336 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/1/real_time_median                 335 us          335 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/1/real_time_stddev                2.36 us         2.35 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/1/real_time_mean          666 us          355 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/1/real_time_median        664 us          355 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/1/real_time_stddev       8.92 us         1.03 us            3
``` 

So some thoughts I have that I have **not** experimented/validated yet. Is this a startup cost assocaited in the instantiation of the Scheduler object? If so should I put it inside of a class variable (created only once) instead of creating the scheduler on the function call. Or is this just the penalty that `marl` has, which is not that *steep* of a penalty.

Note that I would not try to reproduce these results with my repository because you need the input data and I feel its too complex. If this startup cost is unexpected it would be better for me to generate a MRE in my benchmark repo. 

2. There was a previous issue brought up by another [here](https://github.com/google/marl/issues/79#issuecomment-584234615). The point I want to highlight is the use of RAII. I kind of agree that it felt pretty weird/different to use the `defer` macro. I have read the issue and I understand it is optional but I felt there were not alot of example to show how to accomplish without using `defer`. I also at one point was trying to call 2 different `wait group .dones()` in one `defer` and that did not work well (this was before in the poorly written `binary_tree` benchmark for `marl`).  I think the idea is that I'm really not that amazing at C++ and the common idiom of [RAII I seem to get](https://www.programming-books.io/essential/cpp/locking-898c28deb8f54157be572e625d1880e0), but the  `defer` makes me a little puzzled about whats going on and what I need to do to sure that no matter what (even exceptions) the wait group gets decremented so I dont wait forever.

3. The API of `cpp-task-flow`, even though I really haven't used it too much either, seems amazing. I'm really impressed by how succinct and simple it is to describe dependencies between tasks and then execute them.  Its probably out of scope for `marl`, but it would nice to have something similar.

If I think of anything else I will let you know. Thanks again for releasing this product and putting such an effort on improvement.

**Comment by ben-clayton** [comment]
Thank you for your feedback @JeremyBYU.

I'll write a longer reply but I wanted to quickly reply to the question regarding constructing the Scheduler. I added some inline comments to your polylidar changes: https://github.com/JeremyBYU/polylidar/commit/9cf2a187008d1cdf54cfe24f10ee7eb67a1ef0e1#r38447227 and https://github.com/JeremyBYU/polylidar/commit/9cf2a187008d1cdf54cfe24f10ee7eb67a1ef0e1#r38447324. The TL;DR is that building schedulers is expensive and really should be only done once.

I'll try to reply to the rest of your comments soon.

Thanks,
Ben

**Comment by JeremyBYU** [comment]
Thanks for the help and feedback. Please take your time, there really is no rush.

**Comment by ben-clayton** [comment]
Hi @JeremyBYU,

Thank you ever so much for taking the time to write all of this up!

I'll try my best to respond to what you've posted: 

> Many of the tasks for polylidar were dynamic and nested, meaning you have no idea how many subtasks must be completed inside of a task (and all those subtasks must finish before their parent task is completed). It seemed like marl was primarily focused on this scenario.

Correct. However, a marl task is free to schedule any number of child tasks _without_ waiting on them to finish, if you so wish. I can see that for your use case, you typically want your tasks to fan-out then fan-in, which marl also covers with `marl::WaitGroup` (and as you've correctly implemented).

> Also I do not have thousands of tasks generally, maybe like 100 total. I figured the difference between all these libraries might be small at such a low amount of tasks.

I would hope so. As I mentioned on the [original post](https://github.com/JeremyBYU/TaskLibraryBenchmark/issues/1), marl does struggle to perform as well as `cpp-task-flow` and `tbb` in terms of task scheduling / execution overhead (although I'd certainly like to narrow the gap here). That said, I believe marl has unique strengths in its simplistic interface for dynamic task scheduling and synchronization - specifically that it does not need to pre-declare the dependency graph for the tasks. As you point out, once the workload size is moderately beefy, the scheduling overhead becomes far less of an issue.

>  So some thoughts I have that I have not experimented/validated yet. Is this a startup cost assocaited in the instantiation of the Scheduler object? If so should I put it inside of a class variable (created only once) instead of creating the scheduler on the function call. Or is this just the penalty that marl has, which is not that steep of a penalty.

Already partially covered above and [in this thread](https://github.com/JeremyBYU/polylidar/commit/9cf2a187008d1cdf54cfe24f10ee7eb67a1ef0e1#r38447227), but repeating and expanding here for completeness: Yes, constructing a `marl::Scheduler` is expensive, and I recommend only doing it once, per process.

The typical usage of the marl scheduler is to create one instance at app startup (say in main), and destruct it just before the application is terminated. Creating a scheduler is relatively expensive as it's spawning all the worker threads, and internal worker state. Destructing is also expensive, as it has to wait on all submitted work, and then tear down all the worker threads. I also do not recommend having more than one active scheduler per process - they'll fight each other for CPU utilization.
Once you've created a scheduler and called `marl::scheduler::bind()` it is bound to the calling thread via thread-local-storage, and can be picked up anywhere on that thread without parameter passing.

If you want to call `marl::schedule()` from other threads you have created yourself, you'll need to bind the scheduler to those additional threads by acquiring the scheduler with `marl::Scheduler::get()` and calling `marl::Scheduler::bind()` on those other threads. Just remember to unbind the scheduler from each thread before the thread exits, otherwise the scheduler destructor will hang waiting for the unbinds. 

If you're using marl in a library and are not notified of application initialization and termination, you may wish to lazily construct a shared scheduler in your library. [SwiftShader](https://github.com/google/swiftshader) is a library that uses marl, and accomplishes exactly this with a global [`std::weak_ptr<marl::Scheduler>`](https://github.com/google/swiftshader/blob/9d2fd9c161ab6af2cd17c4170e083a63a2a3077f/src/Vulkan/libVulkan.cpp#L134-L151). With this approach we ensure that the scheduler is lazily constructed and destructed on first and last use of [a Vulkan `Device` instance](https://github.com/google/swiftshader/blob/9d2fd9c161ab6af2cd17c4170e083a63a2a3077f/src/Vulkan/libVulkan.cpp#L793-L794), respectively.

As you correctly pointed out, users of your library may be using marl themselves. In this sort of situation, I'd probably encourage your library to expose an API for controlling who constructs and owns the scheduler, but you could just check to see if the caller has already bound an external scheduler with `marl::Scheduler::get()`, and use that instead of creating you own. Be aware though that in this situation, you're probably going to have more issues to do with two copies of marl being linked than anything else. Until marl is more widely used, this is probably all a non-issue.

With all that said, I'd love to know if moving the scheduler construction out of your `ExtractPlanesAndPolygonsOptimized` function fixes this performance concern (or not!).

> The point I want to highlight is the use of RAII. I kind of agree that it felt pretty weird/different to use the defer macro. I have read the issue and I understand it is optional but I felt there were not alot of example to show how to accomplish without using defer. I also at one point was trying to call 2 different wait group `.dones()` in one defer and that did not work well (this was before in the poorly written binary_tree benchmark for marl). 

`defer()` definitely seems to be a hot topic and is something some people strongly dislike. Under the hood it is simply using RAII, but removing the ugly boiler plate by declaring a uniquely named local variable and building the lambda for you. Boost has [`BOOST_SCOPE_EXIT`](https://www.boost.org/doc/libs/1_54_0/libs/scope_exit/doc/html/index.html), which appears to work in a similar vein, but is more clunky to use, IMHO.

You raise a good point about the examples using `defer()`. Perhaps I could rework them to not use `defer()` if the general consensus is that people do not want/like it. For the examples, I could just move the statement to the end of the scope, but this will not deal with accidental early returns or exceptions. 

As for attempting to do multiple `.done()` calls in a single defer - once you realize the macro argument is just fed into a lambda's body, you'll see you can use `;` to delimit multiple statements:

```c++
defer(a.done(); b.done());
```

Is equivalent to:

```c++
defer(b.done());
defer(a.done());
```

Out of curiosity, how did you attempt to combine the two `.done()` calls? Was it like this `defer(a.done(), b.done());`? I could probably make that work, but I sincerely question whether I should.

I do agree that `defer` is all rather magical and is clearly not documented well enough. At the very least, I can try improving the documentation here.

> The API of `cpp-task-flow`, even though I really haven't used it too much either, seems amazing. I'm really impressed by how succinct and simple it is to describe dependencies between tasks and then execute them. Its probably out of scope for marl, but it would nice to have something similar.

Yes, `cpp-task-flow` is an exceptionally well designed API for running pre-defined task graphs. **Not** declaring the dependency graph ahead of time was one of marl's initial design goals, but your benchmarks clearly highlight that we're leaving quite a chunk of performance on the floor for not being able to do this. I also really like the ability to spit out graphviz flow charts, and without this dependency graph, this is near impossible for marl to do right now (although marl does emit traces - not quite the same thing). The idea of layering a similar API on top of marl is an intriguing one, and something I'll mull over for a bit. 

Anyway, I'd love to know how you get on with marl. Please don't hesitate to reach out if you have questions, complaints, feature requests, etc.

Meanwhile, I'll continue thinking of how to close the performance gap in your benchmarks. :)

Many thanks,
Ben

**Comment by JeremyBYU** [comment]
> Out of curiosity, how did you attempt to combine the two .done() calls? Was it like this defer(a.done(), b.done());?

Thats exactly what I tried. I don’t propose you change anything to make this syntax work. I just am not familiar with macro functions and couldn’t actually understand what was going on (I don’t remember exactly but it was like a macro calling another macro and I couldn’t figure out what I needed so I just split into two defers). Now that you explained it if I need to I will just use a semi colon.

Thanks again for the help. I’ll update in the changes we discussed to Polylidar in this thread once they are made. Feel free to close or leave open the issue.

**Comment by JeremyBYU** [comment]
I modified `polylidar` so that the `marl` scheduler is initialized during [class creation](https://github.com/JeremyBYU/polylidar/blob/13c8f14071833c82c17ac6c2c4cdc60dab08ff2d/src/Polylidar/Polylidar.cpp#L27-L34).  I have a member variable now for `std:shared_ptr<marl::Scheduler>`.  Benchmarks indicate much better performance for extremely light workloads (~300 us). There does not seem to be any significant difference between using or not using marl in the smallest workload (has the `/1/` indicator), `324us vs 330 us`. As some *small* opportunities for parallelization  are given (`/2/` -> `/3/` -> `/4/`) marl begins to edge out as well and win. A max of 4 cores are provided to the scheduler in the tests.
```
Benchmark                                                                                               Time             CPU   Aggregate Iterations
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/1/real_time_median                 324 us          324 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/2/real_time_median                 377 us          377 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/3/real_time_median                 580 us          580 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormals/4/real_time_median                 838 us          838 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/1/real_time_median        330 us          214 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/2/real_time_median        391 us          257 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/3/real_time_median        479 us          305 us            3
ImagesAndSparseMesh/BM_ExtractPlanesAndPolygonsFromMultipleNormalsOptimized/4/real_time_median        650 us          388 us            3
```

Thanks again for your help

**Comment by ben-clayton** [comment]
Great to hear that pulling the scheduler construction initialization out helped with performance.

I've [added some usage recommendations](https://github.com/google/marl/commit/748d3c16165c7dcff018273cfaf9d8f26a33333a) to the README.md that includes this, along with other things we have discussed.

**Comment by JeremyBYU** [comment]
Thanks for the update. One question/thought here:

> Calling a non-marl blocking function on a marl worker thread will prevent that worker thread from being able to switch to execute other tasks until the blocking function has returned.

Many of my tasks have a blocking function (high computation, big loops) so I will hit this, but I think that is fine. I was just thinking that this basically means that "task-switching" can only occur when when calling a `wait` on some `marl` primitive.  This means if you don't call `wait` you are basically guaranteed that your function will continue and finish and not worry about a "task switch" interfering the flow. 

If I undetrstand correctly Python has a mechanism where it will automatically switch "threads" (python threads all running on the same cpu core) after a given amount of time, to give other threads the opportunity to do work as well. 
 
https://docs.python.org/3/c-api/init.html#thread-state-and-the-global-interpreter-lock
> In order to emulate concurrency of execution, the interpreter regularly tries to switch threads (see sys.setswitchinterval()).  

I was under the impression that Marl would do that as well. However I'm actually glad that it doesn't because it feels I have more control about task switching.

Thanks.

**Comment by ben-clayton** [comment]
> Many of my tasks have a blocking function (high computation, big loops) so I will hit this

I am referring to calls where no progress can be made until some thread-external signal is raised, such as a network socket waiting on a connection or packet, reading from a slow storage device, or say calling a thread sleep method. In all these cases, the OS will context switch the thread away from your process until that blocking call is complete. That removes a marl worker thread from processing other tasks you have scheduled. 

High computation is good! The marl worker (and so CPU core) is doing something useful.

The only time large work loads may be bad is if these could be split up into smaller workloads so they can be distributed across other idle worker threads. You don't want them too small, otherwise you'll be spending more time in the scheduler than actually doing useful work (as you well know!)

> However I'm actually glad that it doesn't because it feels I have more control about task switching.

Right - workers only ever switch tasks when a *marl* blocking call is made. The OS is always able to context switch the physical core to service other processes (or other threads in your process), but a marl worker will continue executing the current task until it is unable to make forward progress. When a task does become blocked, the worker will perform a fiber switch to save the current state of the blocked task and then start or resume another task.

If you're interested in the gory details, there's a write up of the scheduler [here](https://github.com/google/marl/blob/master/docs/scheduler.md).
